File: all-submissions/0813-all-paths-from-source-to-target/solution.java
class Solution {
    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        List<List<Integer>> answer = new ArrayList<>();
        int length = graph.length;
        List<Integer> curr = new ArrayList<>();
        curr.addLast(0);
        backtrack(0, graph, length, answer, curr);
        return answer;
    }

    public void backtrack(int currNode, int[][] graph, int length, List<List<Integer>> answer, List<Integer> curr) {
        if (currNode == length -1) {
            answer.add(new ArrayList<Integer>(curr));
            return;
        }
        for(int neighbor: graph[currNode]) {
            curr.addLast(neighbor);
            backtrack(neighbor, graph, length, answer,curr);
            curr.removeLast();
        }
    }
}

File: all-submissions/0813-all-paths-from-source-to-target/solution.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        res = []
        def dfs(n, ans):
            if n == len(graph)-1:
                res.append(ans[:])
            for i in graph[n]:
                ans.append(i)
                dfs(i, ans)
                ans.pop()

        dfs(0, [0])
        return res
